### ISSUE TITLE
how to patch ledge the right way?

### ISSUE THREAD

**Issue by keen99** [pintsized/ledge#125]
> Generally, I would try to keep Ledge and all other modules at the latest release, and do anything custom outside of it. For example, your redis timeout issue is fixed in the latest lua-resty-qless and you can update headers by binding to the "before_save" event, in config, rather than changing the source code.

(pulling this out of another thread to help other users, since this isn't really documented)

We needed to add a header to our cache - since we keep our cache for a long time, we wanted to add this to the response that comes out of the cache.   We also add one for new content (important because the -first- request for an object (or an update) wouldn't pass through the cache)

```
-- Add some cache headers.  Ledge will assume they came from the origin and override origin header
ledge:bind("origin_fetched", function(res)
    res.header["X-Origin-Date"] = os.date( "!%Y-%m-%dT%TZ", ngx.parse_http_time(res.header["Date"]) )
end)
```

We achieved this inside ledge.lua:

when pulling data out of the cache, store off the timestamps using the data that the Age header uses:

```
...
        elseif cache_parts[i] == "expires" then
            res.remaining_ttl = tonumber(cache_parts[i + 1]) - ngx_time()
        elseif cache_parts[i] == "saved_ts" then
            time_in_cache = ngx_time() - tonumber(cache_parts[i + 1])
            -- DSR - save date of the original request for use in a header
            saved_ts = tonumber(cache_parts[i + 1])
        elseif cache_parts[i] == "generated_ts" then
            time_since_generated = ngx_time() - tonumber(cache_parts[i + 1])
            -- DSR - save date of the original request for use in a header
            generated_ts = tonumber(cache_parts[i + 1])
        elseif cache_parts[i] == "has_esi" then
...
```

Then, along with the Age header, create our new header:

```
        -- Calculate the Age header
        if res.header["Age"] then
            -- We have end-to-end Age headers, add our time_in_cache.
            res.header["Age"] = tonumber(res.header["Age"]) + time_in_cache
            -- DSR - return ISO-8601 date of the original request
            res.header["X-Origin-Date"] =  os.date( "!%Y-%m-%dT%TZ", tonumber(saved_ts) )
        elseif res.header["Date"] then
            -- We have no advertised Age, use the generated timestamp.
            res.header["Age"] = time_since_generated
            -- DSR - return ISO-8601 date of the original request
            res.header["X-Origin-Date"] =  os.date( "!%Y-%m-%dT%TZ", tonumber(generated_ts) )
        end
```

So this works fine for us - but, is there a way to apply these two patches to _M.read_from_cache() without modifying the code?   If there is - how would someone identify what to "bind" to for a particular section of the code?

We have other patches too:

```
    -- set the host header  - DSRdsr
    -- https://github.com/pintsized/ledge/issues/98
-- alternatively, we could do this in a  bind to the before_request and set req.header["Host"]
    headers['host'] = self:config_get("upstream_host")
```

For that one - it oculd be applied through binding to before_request - but how would I discover that that's the right place to put it (except that github issue)?

What about other components, such as redis/connector.lua?  (where I had to patch a timeout due to a qless bug)   Are those bind()able as well?

thanks.  just trying to get better visibility into this approach.

**Comment by pintsized** [comment]
Hi, and sorry for the insanely slow replies! Things have been a bit hectic.

> We needed to add a header to our cache - since we keep our cache for a long time, we wanted to add this to the response that comes out of the cache. We also add one for new content (important because the -first- request for an object (or an update) wouldn't pass through the cache)

...

> So this works fine for us - but, is there a way to apply these two patches to _M.read_from_cache() without modifying the code? If there is - how would someone identify what to "bind" to for a particular section of the code?

The most commonly useful events are "origin_fetched" for modifying a cache MISS before saving (as you have in your example), and "response_ready" for modifying all request prior to serving them.

In your case, I think you can do all of this in "origin_fetched" - i.e. store your header value on save rather than dynamically creating them on cache read. You're trying to use some internal stored metadata to determine the headers, but you can just as easily reproduce that. `saved_ts` for example is just `ngx.time()` at the point of saving. `generated_ts` is a timestamp derived from the Date header.

I'm not sure I entirely follow what your X-Origin-Date header needs to be, but you can either base it on the current time at the point of save, or on another header value if needed. Either way, "origin_fetched" in your config seems like the place to put this, rather than patching the core code. Which is what you do in your first example - so I'm a bit confused as to what else you need to do?

> For that one - it oculd be applied through binding to before_request - but how would I discover that that's the right place to put it (except that github issue)?

Well, they are documented, admittedly quite minimally: https://github.com/pintsized/ledge#event-types

I'm really hoping to get some time to write significantly better documentation soon, with examples etc. Aside from some of it being a bit thin, I'm pretty sure some stuff is just wrong / out of date too, so this definitely needs looking at.

In summary though, the hope is that you can do everything you need via binding to events. If there's something you can't do, raise an issue as perhaps the answer is a new event or some additional event parameters etc.
